### PR TITLE
[REVIEW] Port hashing to cudf++

### PR DESCRIPTION
**Pull request by trevorsm7**
Part of issue https://github.com/rapidsai/cudf/issues/2931, porting hashing functions from functions.h to cudf++

- [x] Define API
- [x] Implement API
- [x] Implement tests

Merge after:

- [x] https://github.com/rapidsai/cudf/pull/3310
- [x] https://github.com/rapidsai/cudf/pull/3313
- [x] https://github.com/rapidsai/cudf/pull/3371
- [x] https://github.com/rapidsai/cudf/pull/3448

### CODE REVIEW

File: cpp/include/cudf/hashing.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * @param[in] columns_to_hash Indices of the columns in the input set to hash
+ * @param[in] num_partitions The number of partitions to rearrange the input rows into
+ * @param[in] hash The hash function to use
+ * 
+ * @returns A pair of <partioned output, partition offsets>, where the partitioned
+ * output is a new table with rows from the input reordered into partitions and
+ * the partion offsets are a vector of offsets to the first index of each partition
+ * ----------------------------------------------------------------------------**/
+std::pair<std::unique_ptr<table>, std::vector<size_type>>
+hash_partition(table_view const& input,
```
Comment (jrhemstad) [inline cpp/include/cudf/hashing.hpp]: This should probably be updated to return `std::vector<std::unique_ptr<table>>`, i.e., it actually does the partitioning of `input` into discrete tables.

Comment (trevorsm7) [reply cpp/include/cudf/hashing.hpp]: Good point, that's a lot cleaner than the old signature

**Comment by trevorsm7** [comment]
Uses https://github.com/rapidsai/cudf/pull/3310

File: cpp/src/hash/hashing.cu [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+  // Compute in-place exclusive scan of size of each partition to determine
+  // offset location of each partition in final output.
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         global_partition_sizes.begin(),
+                         global_partition_sizes.end(),
+                         global_partition_sizes.begin());
+
+  // Copy the result of the exlusive scan to the output offsets array
+  // to indicate the starting point for each partition in the output
+  std::vector<size_type> partition_offsets(num_partitions);
```
Comment (jrhemstad) [inline cpp/src/hash/hashing.cu]: Use a `thrust::host_vector` for automatic device to host copying.

File: cpp/include/cudf/table/row_operators.cuh [hunk 3]
```diff
@@ -409,7 +409,7 @@ template <template <typename> class hash_function, bool has_nulls = true>
 class row_hasher {
  public:
   row_hasher() = delete;
-  row_hasher(table_device_view t) : _table{t} {}
+  __host__ __device__ row_hasher(table_device_view t) : _table{t} {}
```
[the remark this answers is not in the document]
Comment (trevorsm7) [reply cpp/include/cudf/table/row_operators.cuh]: `row_hasher` is constructed/used inside the `compute_row_partition_numbers` kernel, so I had to make it device callable

**Comment by nsakharnykh** [comment]
@trevorsm7 just want to reference #2623 with some optimizations for hash partitioning. I'm assuming your PR is just updating the current implementation to libcudf++. It might be useful to see if we can also get the optimizations from #2623 ported to the new code in the future.

File: cpp/include/cudf/table/table_view.hpp [hunk 4]
```diff
@@ -141,6 +141,19 @@ class table_view_base {
...
+   * @brief Returns a table_view with set of specified columns.
+   *
+   * @throws cudf::logic_error
+   * If any element in `column_indices` is outside [0, num_columns())
+   *
+   * @param column_indices Indices of columns in the table
+   * @return A table_view consisting of columns from the original table
+   * specified by the elements of `column_indices`
+   *---------------------------------------------------------------------------**/
+  table_view select(std::vector<cudf::size_type> const& column_indices) const;
```
Comment (jrhemstad) [inline cpp/include/cudf/table/table_view.hpp]: @trevorsm7 can you break this into a separate PR so it can be used in https://github.com/rapidsai/cudf/pull/3250?

File: cpp/src/hash/hashing.cu (hunk 2 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Ah good idea. This doesn't take a stream though, so I assuming it's synchronizing on the default stream?

Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Actually... I'm not sure that I need `partition_offsets` in host memory in the new implementation, so this might be a moot point.

File: cpp/tests/hashing/hash_test.cu [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+
+namespace {
+
+// TODO move these to column_utilities?
+
+#define RETURN_FALSE_IF_NEQ(a, b) if (not ((a) == (b))) { return false; }
+
+bool column_properties_equal(cudf::column_view lhs, cudf::column_view rhs) {
```
Comment (trevorsm7) [inline cpp/tests/hashing/hash_test.cu]: Oh shoot, I forgot I was going to remove this. I'll get to it tomorrow.

File: cpp/include/cudf/detail/hashing.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+namespace cudf {
+namespace detail {
+
+/** --------------------------------------------------------------------------*
+ * @brief Partitions a table into the desired number of partitions according to
+ * the hash value of each row. 
+ *
+ * Computes the hash values of the rows in the specified columns of the input
+ * columns and bins the hash values into the desired number of partitions.
+ * Rearranges the input columns such that rows with hash values in the same bin
```
Comment (harrism) [inline cpp/include/cudf/detail/hashing.hpp]: Does it sort them? Or just group by hash value? Maybe this could be more explicit?

```suggestion
 * Sorts the contents of the output columns so that rows with hash values in the same bin
```

File: cpp/src/hash/hashing.cu [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  // Scatter input rows to output partitions given the partition map
+  auto const partition_map = column_view(data_type(INT32), num_rows,
+    row_partition_numbers.data().get());
+  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
+
+  // Pad with empty tables if we have less than num_partitions
+  if (output.size() < static_cast<size_t>(num_partitions)) {
+    auto const old_size = output.size();
+    output.resize(num_partitions);
+    std::generate(output.begin() + old_size, output.end(),
```
Comment (harrism) [inline cpp/src/hash/hashing.cu]: This could be ~simpler~ shorter with `std::back_inserter()`?  Delete lines 116-117 and do this:

It's possible this will reallocate the array more than once, but hopefully the insertion is smart enough to not do that.

```suggestion
    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
```

File: cpp/src/hash/hashing.cu [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+namespace detail {
+
+std::vector<std::unique_ptr<experimental::table>>
+hash_partition(table_view const& input,
+               std::vector<size_type> const& columns_to_hash,
+               int num_partitions,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+{
+  CUDF_EXPECTS(columns_to_hash.size() > 0, "Need at least one column to hash");
```
Comment (harrism) [inline cpp/src/hash/hashing.cu]: Typically in libcudf empty inputs result in empty outputs, not errors. This is important, for example, for clients like Dask (and I think Spark).

File: cpp/src/hash/hashing.cu [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+template <typename hash_value_t>
+struct bitwise_partitioner
+{
+  bitwise_partitioner(size_type num_partitions) : divisor{(num_partitions - 1)} {
+    assert(is_power_two(num_partitions));
+  }
+
+  __host__ __device__
+  size_type operator()(hash_value_t hash_value) const {
+    return hash_value & (divisor);
```
Comment (rgsl888prabhu) [inline cpp/src/hash/hashing.cu]: Should this be?
```suggestion
    return hash_value & (divisor-1);
```

Comment (kaatish) [reply cpp/src/hash/hashing.cu]: Divisor is already of the type 2^n - 1 so I think this is fine. Perhaps the documentation can use explicit variable names to clear this out.

File: cpp/tests/hashing/hash_test.cu [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+class HashTest : public cudf::test::BaseFixture {};
+
+TEST_F(HashTest, MultiValue)
+{
+  using cudf::test::fixed_width_column_wrapper;
+  using cudf::test::strings_column_wrapper;
+  using cudf::experimental::bool8;
+
+  auto const strings_col = strings_column_wrapper(
```
Comment (kaatish) [inline cpp/tests/hashing/hash_test.cu]: This line is giving you a compilation error on ubuntu16.04-cuda9.2

```suggestion
  strings_column_wrapper strings_col({
```

Comment (trevorsm7) [reply cpp/tests/hashing/hash_test.cu]: Good catch! I had this same problem in my scatter tests as well. Seems to be a gcc bug :(

File: cpp/tests/hashing/hash_test.cu (hunk 5 above)
Comment (trevorsm7) [reply cpp/tests/hashing/hash_test.cu]: Done

File: cpp/include/cudf/detail/hashing.hpp (hunk 6 above)
Comment (trevorsm7) [reply cpp/include/cudf/detail/hashing.hpp]: It doesn't actually sort them. How does this sound?
> Partitions rows of `input` into `num_partitions` bins based on the hash
> value of the columns specified by `columns_to_hash`. Rows partitioned into
> the same bin are grouped together into a new table. Returns a vector
> containing `num_partitions` new tables.

File: cpp/src/hash/hashing.cu (hunk 7 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Done

File: cpp/src/hash/hashing.cu (hunk 9 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Ok, I made this more clear with a comment and change of variable name: `return hash_value & mask; // hash_value & (num_partitions - 1)`

File: cpp/src/hash/hashing.cu [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+{
+  std::vector<std::unique_ptr<experimental::table>> output;
+  auto table_to_hash = input.select(columns_to_hash);
+
+  // Return empty vector if there are no partitions or anything to hash
+  if (num_partitions <= 0 || input.num_rows() == 0 || table_to_hash.num_columns() == 0) {
+    return output;
+  }
+
+  cudf::nvtx::range_push("CUDF_HASH_PARTITION", cudf::nvtx::PARTITION_COLOR);
```
Comment (harrism) [inline cpp/src/hash/hashing.cu]: Why not start the range at the start of the function?

File: cpp/src/hash/hashing.cu (hunk 8 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Done

File: cpp/src/hash/hashing.cu (hunk 11 above)
Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: I would need to add corresponding range_pop() calls when input.select() throws an exception and when I return early when the input is empty. If we had an RAII NVTX wrapper that could be automatic! But, I think this is a little cleaner anyway since you would only see the range show up when there's actually work being done.

**Comment by trevorsm7** [comment]
rerun tests

File: cpp/src/hash/hashing.cu (hunk 11 above)
Comment (harrism) [reply cpp/src/hash/hashing.cu]: If there's no work being done the range will be so short you won't see it.  I agree RAII NVTX is needed. It would also allow you to remove the early declaration of output which would make @codereport happy.

Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: Right! Then I could just do this:
`  if (has_nulls(table_to_hash)) {
    return hash_partition_table<true>(
        input, table_to_hash, num_partitions, mr, stream);
  } else {
    return hash_partition_table<false>(
        input, table_to_hash, num_partitions, mr, stream);
  }`

Comment (trevorsm7) [reply cpp/src/hash/hashing.cu]: I think @jrhemstad had mentioned that he had an RAII wrapper for NVTX already?

Comment (jrhemstad) [reply cpp/src/hash/hashing.cu]: > If we had an RAII NVTX wrapper that could be automatic!

Yeah, it's on my backburner:
https://github.com/rapidsai/cudf/issues/2397
https://github.com/jrhemstad/nvtx_wrappers

**Comment by mike-wendt** [comment]
rerun tests

**Comment by harrism** [comment]
rerun tests

**Comment by jrhemstad** [comment]
rerun tests
